### PR TITLE
Tozpatron 51 image controller

### PR DESCRIPTION
**Pull request by macieg-b**
Add controller enables to upload file. Returns URL to uploaded file.

### CODE REVIEW

File: src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<String> uploadFile(@RequestParam("file") MultipartFile file) {
+        validateImageArgument(file);
+        final UploadedFile uploadedFile = storageService.store(file);
+        final URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+                .build().toUri();
+        final String imageUrl = String.format("%s/%s/%s", location.toString().replace("/images", ""), this.storageProperties.getStoragePathRoot(), uploadedFile.getPath());
+
+        return ResponseEntity.created(location)
+                .body(imageUrl);
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java]: Controller states, it's producing JSON, but POST method is returning string.

File: src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @ApiOperation("Upload image")
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Bad Request"),
+            @ApiResponse(code = 422, message = "Invalid image file")
+    })
+
+    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<UploadedFile> uploadFile(@RequestParam("file") MultipartFile file) {
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java]: Images should be uploaded in pet context.

File: src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        validateImageArgument(file);
+        final UploadedFile uploadedFile = storageService.store(file);
+        final URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+                .build().toUri();
+        return ResponseEntity.created(location)
+                .body(uploadedFile);
+
+    }
+
+    private void validateImageArgument(MultipartFile multipartFile) {
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java]: Method name should be simply validateImage or validateImageFile

File: src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        try (InputStream input = multipartFile.getInputStream()) {
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/ImageUploadController.java]: Simplier:
```
try (InputStream input = multipartFile.getInputStream()) {
    ImageIO.read(input).toString();
} catch(Exception e) {
    throw new InvalidImageFileException(IMAGES);
}
```

File: src/main/java/com/intive/patronage/toz/exception/InvalidImageFileException.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.intive.patronage.toz.exception;
+
+public class InvalidImageFileException extends RuntimeException {
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/exception/InvalidImageFileException.java]: Maybe some base / common class for client exception?

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java [hunk 6]
```diff
@@ -83,4 +108,31 @@ public PetView getPetById(@ApiParam(required = true) @PathVariable UUID id) {
     public PetView updatePet(@PathVariable UUID id, @RequestBody PetRequestBody pet) {
         return petsService.updatePet(id, pet);
     }
+
+    @ApiOperation("Upload image")
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Bad Request"),
+            @ApiResponse(code = 422, message = "Invalid image file")
+    })
+    @PostMapping(value = "/images/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: So, we will have separate controllers for uploading images for pets and for uploading images for news, but they will use the same service, right?

Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: According to our conversation.

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java [hunk 7]
```diff
@@ -83,4 +108,31 @@ public PetView getPetById(@ApiParam(required = true) @PathVariable UUID id) {
...
+        petsService.updatePetImage(id, uploadedFile.getPath());
+        final URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+                .build().toUri();
+        UrlView urlView = new UrlView();
+        urlView.setUrl(String.format("%s/%s/%s", location.toString().replace("/images", ""), this.storageProperties.getStoragePathRoot(), uploadedFile.getPath()));
+        return ResponseEntity.created(location)
+                .body(urlView);
+    }
+
+    private void validateImageArgument(MultipartFile multipartFile) {
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: If other controllers will be using file uploading, then this method also should be copied there. Wouldn't it be better to move this to service?

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java [hunk 8]
```diff
@@ -83,4 +108,31 @@ public PetView getPetById(@ApiParam(required = true) @PathVariable UUID id) {
...
+        try (InputStream input = multipartFile.getInputStream()) {
+            if (ImageIO.read(input) == null)
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: Use braces, even for one liners.

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java [hunk 9]
```diff
@@ -83,4 +108,31 @@ public PetView getPetById(@ApiParam(required = true) @PathVariable UUID id) {
...
+                throw new InvalidImageFileException("Images");
+        } catch (IOException e) {
+            throw new InvalidImageFileException("Images");
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: We lost here real cause of exception here. You should add it to `InvalidImageFileException`.

File: src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java [hunk 10]
```diff
@@ -96,4 +97,16 @@ public ErrorResponse handleWrongEnumValueException(WrongEnumValueException e) {
         final HttpStatus httpStatus = HttpStatus.UNPROCESSABLE_ENTITY;
         return new ErrorResponse(httpStatus.value(), httpStatus.getReasonPhrase(), message);
     }
+
+    @ExceptionHandler(InvalidImageFileException.class)
+    @ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY)
+    @ResponseBody
+    public ErrorResponse invalidImageFileException(InvalidImageFileException e) {
+        String message = messageSource.getMessage("unprocessableEntity",
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java]: Here we should log and maybe even return real exception message (root cause) to the client. You could reuse our error model for that - validation should work.

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java [hunk 11]
```diff
@@ -83,4 +108,31 @@ public PetView getPetById(@ApiParam(required = true) @PathVariable UUID id) {
...
+    })
+    @PostMapping(value = "/images/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<UrlView> uploadFile(@PathVariable UUID id, @RequestParam("file") MultipartFile file) {
+        validateImageArgument(file);
+        final UploadedFile uploadedFile = storageService.store(file);
+        petsService.updatePetImage(id, uploadedFile.getPath());
+        final URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+                .build().toUri();
+        UrlView urlView = new UrlView();
+        urlView.setUrl(String.format("%s/%s/%s", location.toString().replace("/images", ""), this.storageProperties.getStoragePathRoot(), uploadedFile.getPath()));
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: Why this is hardcoded: `/images`? Maybe you don't need to use `location` at all and return `/storageRoot/path/to/image` instead?

File: src/main/java/com/intive/patronage/toz/model/view/UrlView.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class UrlView {
+
+    @ApiModelProperty(example = "http://localhost.com", position = 1)
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/view/UrlView.java]: Wrong `example` it should be `/storageRoot/path/to/image`.

File: src/test/java/com/intive/patronage/toz/controller/PetsControllerImagesTest.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(SpringRunner.class)
+@AutoConfigureMockMvc
+@SpringBootTest
+public class PetsControllerImagesTest {
+
+    private final static String PATH = "/images";
```
Comment (mlotysz) [inline src/test/java/com/intive/patronage/toz/controller/PetsControllerImagesTest.java]: What is PATH? Name it by its function. I hope it's `STORAGE_DIR` or `STORAGE_ROOT`.

File: src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java [hunk 14]
```diff
@@ -85,6 +85,17 @@ public ErrorResponse handleNotFoundException(NotFoundException e) {
...
+    @ExceptionHandler(WrongEnumValueException.class)
+    @ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY)
+    @ResponseBody
+    public ErrorResponse handleWrongEnumValueException(WrongEnumValueException e) {
+        final String typeName = e.getName();
+        final String allowedValues = StringUtils.join(Arrays.asList(e.getAllowedValues()), ", ");
+        final String message = messageSource.getMessage("mustHaveValue",
+                new String[]{typeName, allowedValues},
+                LocaleContextHolder.getLocale());
+        final HttpStatus httpStatus = HttpStatus.UNPROCESSABLE_ENTITY;
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java]: Don't create local variables if they aren't needed.

File: src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java [hunk 15]
```diff
@@ -85,6 +85,17 @@ public ErrorResponse handleNotFoundException(NotFoundException e) {
         return new ErrorResponse(HttpStatus.NOT_FOUND.value(), HttpStatus.NOT_FOUND.getReasonPhrase(), message);
     }
 
+    @ExceptionHandler(WrongEnumValueException.class)
+    @ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY)
+    @ResponseBody
+    public ErrorResponse handleWrongEnumValueException(WrongEnumValueException e) {
+        final String typeName = e.getName();
+        final String allowedValues = StringUtils.join(Arrays.asList(e.getAllowedValues()), ", ");
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java]: Maybe this logic should be in WrongEnumValueException?

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java (hunk 6 above)
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: I guess that id is a pet identifier? If yes the path is modeled incorrectly.
```/pets/{petId}/gallery/```
looks better

File: src/main/java/com/intive/patronage/toz/model/db/Pet.java [hunk 16]
```diff
@@ -20,6 +20,7 @@
     private PetValues.Sex sex;
     private String description;
     private String address;
+    private String imageUrl;
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/model/db/Pet.java]: Are we limit gallery to single image per pet?

File: src/main/java/com/intive/patronage/toz/service/PetsService.java [hunk 17]
```diff
@@ -100,4 +100,11 @@ private PetView convertToView(final Pet pet) {
         petView.setLastModified(pet.getLastModified());
         return petView;
     }
+
+    public void updatePetImage(final UUID id, String imageUrl) {
+        throwNotFoundExceptionIfNotExists(id);
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/service/PetsService.java]: Method name could be better.

File: src/test/java/com/intive/patronage/toz/controller/PetsControllerImagesTest.java [hunk 18]
```diff
@@ -1,74 +1,102 @@
...
     private StorageService storageService;
+    @MockBean
+    private PetsService petsService;
+
+
+    @Before
+    public void setUp() {
+        pet.setId(UUID.randomUUID());
+
+        UploadedFile uploadedFile = new UploadedFile();
```
Comment (pskrzypek84) [inline src/test/java/com/intive/patronage/toz/controller/PetsControllerImagesTest.java]: UploadedFile preparation should be extracted.

File: src/main/java/com/intive/patronage/toz/service/PetsService.java (hunk 17 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/service/PetsService.java]: updatePetImageUrl() ?

File: src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java (hunk 10 above)
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/ControllerExceptionHandler.java]: `e.getName()` isn't a real cause. You need to get it from `e.getMessage()` and `InvalidImageFileException` should set it from `IOException` you wrap...

File: src/main/java/com/intive/patronage/toz/controller/PetsController.java (hunk 7 above)
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/PetsController.java]: Move it to common place could be service. Next use will show us if that is right decision or not.

File: src/main/resources/application.properties [hunk 19]
```diff
@@ -11,3 +11,5 @@ spring.messages.basename=i18n/messages
 spring.messages.encoding=UTF-8
 
 spring.profiles.active=dev
+spring.http.multipart.max-file-size=16384KB
```
Comment (mlotysz) [inline src/main/resources/application.properties]: Check jira and my comment on slack. AFAIR it should be 20MB for file and request is 2*20MB.

**Pull request merged**